File: 정종길/프로그래머스/주차 요금 계산.java
import java.util.*;

class Solution {
    public ArrayList<Integer> solution(int[] fees, String[] records) {
        ArrayList<Integer> answer = new ArrayList<>();
        TreeMap<String, List<String>> table = new TreeMap<>();

        for(String record: records){
            String[] rec = record.split(" ");
            String time = rec[0];
            String car = rec[1];

            if(table.containsKey(car)){
                table.get(car).add(time);

            }else{
                List<String> tmp = new ArrayList<>();
                tmp.add(time);
                table.put(car, tmp);
            }

        }

        for(String key : table.keySet()){
            int fee = cal_fee(table.get(key), fees, fees[0]);
            answer.add(fee);
        }

        return answer;
    }

    public int cal_fee(List<String> times, int[] fees, int basic) {

        int fee = 0;
        int time = 0;

        if(times.size() % 2 == 1){
            int out = 1439;
            int in = Integer.parseInt(times.get(times.size() - 1).substring(0, 2)) * 60 + Integer.parseInt(times.get(times.size() - 1).substring(3));
            time += out - in;
        }

        for(int i = 0; i < times.size() - 1; i += 2){
            int out = Integer.parseInt(times.get(i + 1).substring(0, 2)) * 60 + Integer.parseInt(times.get(i + 1).substring(3));
            int in = Integer.parseInt(times.get(i).substring(0, 2)) * 60 + Integer.parseInt(times.get(i).substring(3));
            time += out - in;
        }

        if(time - fees[0] > 0){
            fee += fees[1] + Math.ceil(((time - fees[0]) / (double)fees[2])) * fees[3];
        } else {
            fee += fees[1];
        }

        return fee;

    }
}

File: 정종길/프로그래머스/카카오 2회독/2022 KAKAO BLIND RECRUITMENT/주차 요금 계산.py
from collections import defaultdict

def solution(fees, records):
    answer = []
    infos = defaultdict(list)
    
    for record in records:
        time, car, _ = record.split()
        infos[car].append(to_min(time))
    
    
    for key in sorted(infos.keys()):
        infos[key].sort()
        tot = cal_parking(infos[key])
        answer.append(cal_fee(tot, fees))
        
    return answer

def to_min(time):
    return int(time[:2]) * 60 + int(time[3:])

def cal_fee(tot, fees):
    fee = fees[1]
    
    if tot <= fees[0]: return fee
    
    q, r = divmod(tot - fees[0], fees[2])
    fee += (q * fees[3]) if r == 0 else ((q + 1) * fees[3])
    return fee
    
def cal_parking(info):
    t = 0
    if len(info) % 2 == 1:
        info.append(23*60 + 59)
        
    for i in range(0, len(info), 2):
        t += info[i + 1] - info[i]
    
    return t
        